File: MyButler/app/models/screening_models.py
# -*- coding: utf-8 -*-
"""
Screening Models
주식 스크리닝 데이터 모델
"""
from datetime import date, datetime
from typing import List, Optional, Dict, Any
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, Field


class MarketType(str, Enum):
    """시장 유형"""
    US = "US"
    KR = "KR"
    ALL = "ALL"


class SignalType(str, Enum):
    """신호 유형"""
    STRONG_BUY = "STRONG_BUY"
    BUY = "BUY"
    WEAK_BUY = "WEAK_BUY"
    NEUTRAL = "NEUTRAL"
    WEAK_SELL = "WEAK_SELL"
    SELL = "SELL"
    STRONG_SELL = "STRONG_SELL"


class FilterType(str, Enum):
    """필터 유형"""
    ICHIMOKU = "ichimoku"
    BOLLINGER = "bollinger"
    MA_ALIGNMENT = "ma_alignment"
    CUP_HANDLE = "cup_handle"


class CombineMode(str, Enum):
    """필터 조합 모드"""
    ANY = "any"  # OR: 하나라도 충족
    ALL = "all"  # AND: 모두 충족


class ScreeningRequest(BaseModel):
    """스크리닝 요청"""
    market: MarketType = Field(default=MarketType.ALL, description="대상 시장")
    min_score: int = Field(default=50, ge=-100, le=100, description="최소 점수")
    perfect_only: bool = Field(default=False, description="완벽 조건만")
    limit: int = Field(default=20, le=100, description="결과 개수")
    filters: List[str] = Field(
        default=["ichimoku"],
        description="적용할 필터 목록: ichimoku, bollinger, ma_alignment, cup_handle"
    )
    combine_mode: CombineMode = Field(
        default=CombineMode.ANY,
        description="필터 조합 모드: any (OR) 또는 all (AND)"
    )


class StockSignal(BaseModel):
    """종목 신호"""
    ticker: str
    name: str
    market: str
    current_price: float
    signal_strength: str
    score: int

    # 일목균형표 조건 충족 여부
    price_above_cloud: bool = False
    tenkan_above_kijun: bool = False
    chikou_above_price: bool = False
    cloud_bullish: bool = False
    cloud_breakout: bool = False
    golden_cross: bool = False
    thin_cloud: bool = False

    # 일목균형표 수치
    tenkan_sen: float = 0.0
    kijun_sen: float = 0.0
    senkou_span_a: float = 0.0
    senkou_span_b: float = 0.0

    # 거래대금
    avg_trading_value: float = 0.0

    # 새 필터 신호 (볼린저 밴드)
    bollinger_squeeze: bool = False
    bollinger_score: int = 0
    bollinger_bandwidth: Optional[float] = None
    bollinger_percent_b: Optional[float] = None

    # 새 필터 신호 (이동평균 정배열)
    ma_perfect_alignment: bool = False
    ma_alignment_score: int = 0
    ma_disparity: Optional[float] = None

    # 새 필터 신호 (컵앤핸들)
    cup_handle_pattern: bool = False
    cup_handle_score: int = 0
    cup_handle_breakout_imminent: bool = False

    # 보너스 점수
    bonus_score: int = 0
    total_technical_score: int = 0
    active_patterns: List[str] = []

    class Config:
        from_attributes = True


class ScreeningResponse(BaseModel):
    """스크리닝 응답"""
    screening_date: date
    market: str
    total_scanned: int
    total_passed_filter: int
    total_signals: int

    # 신호 목록
    strong_buy: List[StockSignal] = []
    buy: List[StockSignal] = []
    weak_buy: List[StockSignal] = []

    # 요약
    summary: Dict[str, Any] = {}


class ScreeningResultCreate(BaseModel):
    """스크리닝 결과 생성 모델 (DB 저장용)"""
    screening_date: date
    ticker: str
    name: Optional[str] = None
    market: str
    current_price: float
    signal_strength: str
    score: int
    price_above_cloud: bool
    tenkan_above_kijun: bool
    chikou_above_price: bool
    cloud_bullish: bool
    cloud_breakout: bool
    golden_cross: bool
    avg_trading_value: float


class ScreeningResult(ScreeningResultCreate):
    """스크리닝 결과 조회 모델"""
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


class ScreeningHistoryRequest(BaseModel):
    """스크리닝 히스토리 요청"""
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    market: Optional[str] = None
    ticker: Optional[str] = None
    min_score: int = Field(default=50, description="최소 점수")
    limit: int = Field(default=100, le=500)
    offset: int = Field(default=0, ge=0)


class ScreeningHistoryResponse(BaseModel):
    """스크리닝 히스토리 응답"""
    records: List[ScreeningResult]
    total_count: int
    limit: int
    offset: int


class ScreeningStatusResponse(BaseModel):
    """스크리닝 상태 응답"""
    is_running: bool
    last_screening_date: Optional[date] = None
    last_screening_market: Optional[str] = None
    last_total_signals: int = 0
    next_scheduled: Optional[datetime] = None


class DailyRecommendation(BaseModel):
    """일일 추천 종목"""
    date: date
    us_recommendations: List[StockSignal]
    kr_recommendations: List[StockSignal]
    total_count: int
    generated_at: datetime

File: my_butler_web/lib/data/models/screening_models.dart
/// Screening API 데이터 모델
library;

enum MarketType { US, KR, ALL }

enum SignalStrength {
  strongBuy,
  buy,
  weakBuy,
  neutral,
  weakSell,
  sell,
  strongSell;

  static SignalStrength fromString(String value) {
    switch (value.toUpperCase()) {
      case 'STRONG_BUY':
        return SignalStrength.strongBuy;
      case 'BUY':
        return SignalStrength.buy;
      case 'WEAK_BUY':
        return SignalStrength.weakBuy;
      case 'NEUTRAL':
        return SignalStrength.neutral;
      case 'WEAK_SELL':
        return SignalStrength.weakSell;
      case 'SELL':
        return SignalStrength.sell;
      case 'STRONG_SELL':
        return SignalStrength.strongSell;
      default:
        return SignalStrength.neutral;
    }
  }

  String get displayName {
    switch (this) {
      case SignalStrength.strongBuy:
        return '강력 매수';
      case SignalStrength.buy:
        return '매수';
      case SignalStrength.weakBuy:
        return '약매수';
      case SignalStrength.neutral:
        return '중립';
      case SignalStrength.weakSell:
        return '약매도';
      case SignalStrength.sell:
        return '매도';
      case SignalStrength.strongSell:
        return '강력 매도';
    }
  }
}

enum FilterType {
  ichimoku('ichimoku', '일목균형표'),
  bollinger('bollinger', '볼린저밴드'),
  maAlignment('ma_alignment', '이평선정배열'),
  cupHandle('cup_handle', '컵앤핸들'),
  roe('roe', 'ROE'),
  gpm('gpm', 'GPM'),
  debt('debt', '부채비율'),
  capex('capex', 'CapEx');

  const FilterType(this.value, this.displayName);
  final String value;
  final String displayName;
}

enum CombineMode {
  any('any', 'OR'),
  all('all', 'AND');

  const CombineMode(this.value, this.displayName);
  final String value;
  final String displayName;
}

class StockSignal {
  final String ticker;
  final String name;
  final String market;
  final double currentPrice;
  final String signalStrength;
  final int score;

  // 일목균형표
  final bool priceAboveCloud;
  final bool tenkanAboveKijun;
  final bool chikouAbovePrice;
  final bool cloudBullish;
  final bool cloudBreakout;
  final bool goldenCross;
  final bool thinCloud;

  // 볼린저밴드
  final bool bollingerSqueeze;
  final int bollingerScore;
  final double? bollingerBandwidth;
  final double? bollingerPercentB;

  // 이동평균 정배열
  final bool maPerfectAlignment;
  final int maAlignmentScore;
  final double? maDisparity;

  // 컵앤핸들
  final bool cupHandlePattern;
  final int cupHandleScore;
  final bool cupHandleBreakoutImminent;

  // 펀더멘탈
  final int roeScore;
  final double? roeValue;
  final int gpmScore;
  final double? gpmValue;
  final int debtScore;
  final double? debtRatio;
  final int capexScore;
  final double? capexRatio;

  // 종합
  final int totalTechnicalScore;
  final int totalFundamentalScore;
  final List<String> activePatterns;
  final List<String> fundamentalPatterns;

  StockSignal({
    required this.ticker,
    required this.name,
    required this.market,
    required this.currentPrice,
    required this.signalStrength,
    required this.score,
    this.priceAboveCloud = false,
    this.tenkanAboveKijun = false,
    this.chikouAbovePrice = false,
    this.cloudBullish = false,
    this.cloudBreakout = false,
    this.goldenCross = false,
    this.thinCloud = false,
    this.bollingerSqueeze = false,
    this.bollingerScore = 0,
    this.bollingerBandwidth,
    this.bollingerPercentB,
    this.maPerfectAlignment = false,
    this.maAlignmentScore = 0,
    this.maDisparity,
    this.cupHandlePattern = false,
    this.cupHandleScore = 0,
    this.cupHandleBreakoutImminent = false,
    this.roeScore = 0,
    this.roeValue,
    this.gpmScore = 0,
    this.gpmValue,
    this.debtScore = 0,
    this.debtRatio,
    this.capexScore = 0,
    this.capexRatio,
    this.totalTechnicalScore = 0,
    this.totalFundamentalScore = 0,
    this.activePatterns = const [],
    this.fundamentalPatterns = const [],
  });

  SignalStrength get signalType => SignalStrength.fromString(signalStrength);

  factory StockSignal.fromJson(Map<String, dynamic> json) {
    return StockSignal(
      ticker: json['ticker'],
      name: json['name'],
      market: json['market'],
      currentPrice: json['current_price'].toDouble(),
      signalStrength: json['signal_strength'],
      score: json['score'],
      priceAboveCloud: json['price_above_cloud'] ?? false,
      tenkanAboveKijun: json['tenkan_above_kijun'] ?? false,
      chikouAbovePrice: json['chikou_above_price'] ?? false,
      cloudBullish: json['cloud_bullish'] ?? false,
      cloudBreakout: json['cloud_breakout'] ?? false,
      goldenCross: json['golden_cross'] ?? false,
      thinCloud: json['thin_cloud'] ?? false,
      bollingerSqueeze: json['bollinger_squeeze'] ?? false,
      bollingerScore: json['bollinger_score'] ?? 0,
      bollingerBandwidth: json['bollinger_bandwidth']?.toDouble(),
      bollingerPercentB: json['bollinger_percent_b']?.toDouble(),
      maPerfectAlignment: json['ma_perfect_alignment'] ?? false,
      maAlignmentScore: json['ma_alignment_score'] ?? 0,
      maDisparity: json['ma_disparity']?.toDouble(),
      cupHandlePattern: json['cup_handle_pattern'] ?? false,
      cupHandleScore: json['cup_handle_score'] ?? 0,
      cupHandleBreakoutImminent: json['cup_handle_breakout_imminent'] ?? false,
      roeScore: json['roe_score'] ?? 0,
      roeValue: json['roe_value']?.toDouble(),
      gpmScore: json['gpm_score'] ?? 0,
      gpmValue: json['gpm_value']?.toDouble(),
      debtScore: json['debt_score'] ?? 0,
      debtRatio: json['debt_ratio']?.toDouble(),
      capexScore: json['capex_score'] ?? 0,
      capexRatio: json['capex_ratio']?.toDouble(),
      totalTechnicalScore: json['total_technical_score'] ?? 0,
      totalFundamentalScore: json['total_fundamental_score'] ?? 0,
      activePatterns: List<String>.from(json['active_patterns'] ?? []),
      fundamentalPatterns: List<String>.from(
        json['fundamental_patterns'] ?? [],
      ),
    );
  }
}

class ScreeningResponse {
  final DateTime screeningDate;
  final String market;
  final int totalScanned;
  final int totalPassedFilter;
  final int totalSignals;
  final List<StockSignal> strongBuy;
  final List<StockSignal> buy;
  final List<StockSignal> weakBuy;
  final Map<String, dynamic> summary;

  ScreeningResponse({
    required this.screeningDate,
    required this.market,
    required this.totalScanned,
    required this.totalPassedFilter,
    required this.totalSignals,
    required this.strongBuy,
    required this.buy,
    required this.weakBuy,
    required this.summary,
  });

  List<StockSignal> get allSignals => [...strongBuy, ...buy, ...weakBuy];

  factory ScreeningResponse.fromJson(Map<String, dynamic> json) {
    return ScreeningResponse(
      screeningDate: DateTime.parse(json['screening_date']),
      market: json['market'],
      totalScanned: json['total_scanned'],
      totalPassedFilter: json['total_passed_filter'],
      totalSignals: json['total_signals'],
      strongBuy: (json['strong_buy'] as List)
          .map((e) => StockSignal.fromJson(e))
          .toList(),
      buy: (json['buy'] as List).map((e) => StockSignal.fromJson(e)).toList(),
      weakBuy: (json['weak_buy'] as List)
          .map((e) => StockSignal.fromJson(e))
          .toList(),
      summary: json['summary'] ?? {},
    );
  }
}

class DailyRecommendation {
  final DateTime date;
  final List<StockSignal> usRecommendations;
  final List<StockSignal> krRecommendations;
  final int totalCount;
  final DateTime generatedAt;

  DailyRecommendation({
    required this.date,
    required this.usRecommendations,
    required this.krRecommendations,
    required this.totalCount,
    required this.generatedAt,
  });

  factory DailyRecommendation.fromJson(Map<String, dynamic> json) {
    return DailyRecommendation(
      date: DateTime.parse(json['date']),
      usRecommendations: (json['us_recommendations'] as List)
          .map((e) => StockSignal.fromJson(e))
          .toList(),
      krRecommendations: (json['kr_recommendations'] as List)
          .map((e) => StockSignal.fromJson(e))
          .toList(),
      totalCount: json['total_count'],
      generatedAt: DateTime.parse(json['generated_at']),
    );
  }
}

class ScreeningRequest {
  final String market;
  final int minScore;
  final bool perfectOnly;
  final int limit;
  final List<String> filters;
  final String combineMode;

  ScreeningRequest({
    this.market = 'ALL',
    this.minScore = 50,
    this.perfectOnly = false,
    this.limit = 20,
    this.filters = const ['ichimoku'],
    this.combineMode = 'any',
  });

  Map<String, dynamic> toJson() {
    return {
      'market': market,
      'min_score': minScore,
      'perfect_only': perfectOnly,
      'limit': limit,
      'filters': filters,
      'combine_mode': combineMode,
    };
  }
}
